src/core/or/circuitpadding.c

```diff
@@ -253,13 +233,24 @@ circpad_circuit_should_be_marked_for_close(circuit_t *circ, int reason)
                  CIRCUIT_IS_ORIGIN(circ) ?
                  TO_ORIGIN_CIRCUIT(circ)->global_identifier : 0,
                  circuit_purpose_to_string(circ->purpose));
-        return 1;
+
+        return 0; // abort timer reached; mark the circuit for close now
       }
-      return 0;
+
+      /* If we weren't marked dirty yet, let's pretend we're dirty now, to use
+       * that timer instead of the idle one (we're not supposed to look idle,
+       * after all) */
+      if (!circ->timestamp_dirty)
```

Comment:
I really like this commit. But what does this timestamp_dirty thing does here? I was not expecting to see it and I'm not sure what `timestamp_dirty` does in general, or why we need to mod it. Perhaps some more comments?

Reply:
Ah I guess it's so that `circuit_expire_old_circuits_clientside()` definitely applies.

I think we should find a place and write a detailed analysis of how we think that this "block mark for close, but close in the expiry function" codepath should work exactly.

Reply:
More comments: f6d815f7c54f27a3c15f9536430ccb2b1b449dd6

Reply:
More comments: f6d815f7c54f27a3c15f9536430ccb2b1b449dd6